ENH: add plugin system (#148)
exasim-project/FoamAdapter: doc/development/pluginsystem.rst

Plugin System
=============

Modern scientific and engineering workflows require flexible simulation frameworks that can be easily extended and customized.
NeoFOAM's plugin architecture is designed to enable users and developers to add new physics models, boundary conditions, and solver modules without modifying the core codebase.
This approach promotes maintainability, collaboration, and rapid prototyping of new features.

Overview
--------

> The PluginSystem is a runtime-extensible configuration system built on Pydantic discriminated unions and a registry pattern.

Reviewer: ```suggestion
The plugin system is a runtime-extensible configuration system built on Pydantic discriminated unions and a registry pattern.
```

A discriminated union selects the right config class based on a type/tag field (e.g., "shape_type": "circle" vs "shape_type": "square"), and each base class has its own registry of registered child implementations.
All plugins and all models are registered in a central registry to enable easy access and management for UI,validation purposes or generative AI.

> The user would be able to retrieve all available plugins: turbulence models, boundary conditions, etc. and their configuration options and validate them.

Reviewer: ```suggestion
The user can retrieve all available plugins, like turbulence models, boundary conditions, etc.  their configuration options and is able to validate them.
```

> The system also supports the generation of JSON schemas for documentation and validation purposes.

Reviewer: ```suggestion
Additionally, the plugin system supports the generation of JSON schemas for documentation and validation purposes.
```

Registration Process
~~~~~~~~~~~~~~~~~~~~

> The decorator-based registration follows a two-step process:

Reviewer: ```suggestion
The  registration process for the plugins follows a two-step process using decorators:
```

Plugin System
=============

Modern scientific and engineering workflows require flexible simulation frameworks that can be easily extended and customized.

> FoamAdapter's plugin architecture is designed to enable users and developers to add new physics models, boundary conditions, and solver modules without modifying the core codebase.

Reviewer: ```suggestion
NeoFOAM's plugin architecture is designed to enable users and developers to add new physics models, boundary conditions, and solver modules without modifying the core codebase.
```

This approach promotes maintainability, collaboration, and rapid prototyping of new features.

Overview
--------

> The PluginSystem provides a runtime-extensible configuration system using Pydantic discriminated unions and a registry pattern.

Reviewer: ```suggestion
The plugin system provides a runtime-extensible configuration system using Pydantic discriminated unions and a registry pattern.
```

The system allows to register child classes on numerous base classes, each with its own registry.

Core Components
---------------

PluginRegistry Dataclass
~~~~~~~~~~~~~~~~~~~~~~~~

The ``PluginRegistry`` dataclass stores metadata for each plugin base type:

* ``base_cls``: The plugin base class (Pydantic BaseModel)

> * ``discriminator_variable``: Field name holding the union (e.g., 'shape', 'plugin')

Reviewer: Maybe instead of telling what it is, describe what it is required for/used for.

> * ``discriminator``: Discriminator field name in plugin configs (e.g., 'shape_type', 'plugin_type')

Reviewer: Please elaborate what discriminator means.

> * ``plugin_registry``: List of registered plugin configuration classes

Reviewer: needs an explanation what plugin configuration classes are.

> * ``plugin_model``: Dynamically generated extensible Pydantic model

Reviewer: needs more context.

PluginSystem Class
~~~~~~~~~~~~~~~~~~

> The central registry stores all plugin families within the class dictionary variable ``_registry`` .

Reviewer: Needs more elaboration:
- what is "the" central registry
- what are `plugin families`
- `class dictionary variable` -> member variable `_registry` of ? which is a dictionary type

> Each key represents a plugin family name mapping to its corresponding PluginRegistry instance.

Reviewer: I have no idea, what this means to say?

Implementation Details
----------------------

Registration Process
~~~~~~~~~~~~~~~~~~~~

> The decorator-based registration follows a two-step process:

Reviewer: Registration of X in Y uses a two-step process. For this the following  decorators are provided:

> 1. **Base Class Registration**: ``@PluginSystem.register()`` creates a registry entry and adds helper methods

Reviewer: `creates a registry entry` may say where it the entry is created? and what are the helper methods.

> 2. **Plugin Registration**: ``@BaseClass.register`` adds plugin classes to the registry and regenerates the union model

Reviewer: >  regenerates the union model

What does that mean?

Dynamic Model Generation
~~~~~~~~~~~~~~~~~~~~~~~~

> The system uses ``pydantic.create_model()`` to dynamically generate extensible models.

Reviewer: - Pydantic is mentioned quite a lot. Maybe a sentence or two about pydantic would help.
- What is the `The system` ?
- `generate extensible models` -> models that can be extended at runtime?

> Each registration updates the discriminated union type and recreates the model.

Reviewer: `Each registration` of what ? -> Each registration of a new model in the central registry?
- `recreates the model. ` What does recreating the model mean?

> The discriminator field enables automatic deserialization based on the type identifier.

Reviewer: > discriminator field

The `discriminator` field member in the PluginRegistry class 

> enables automatic deserialization based on the type identifier.

what is automatic deserialization?

Usage Examples
--------------

Basic Plugin Setup
~~~~~~~~~~~~~~~~~~

> The following example demonstrates how to define a plugin base class and register multiple plugin configurations.

Reviewer: >  to define a plugin base class  

the plugin base class `ShapeInterface` by inheriting from BaseModel? and register multiple plugins? via the `@SHapeInterface.register ` decorator.

> multiple plugin configurations.

what is a plugin configuration?

.. code-block:: python

    from pydantic import BaseModel
    from typing import Literal
    from foamadapter.core.plugin_system import PluginSystem

    @PluginSystem.register(discriminator_variable="shape", discriminator="shape_type")
    class ShapeInterface(BaseModel):
        color: str
        name: str = "default"

    @ShapeInterface.register
    class CircleConfig(BaseModel):
        shape_type: Literal["circle"]
        radius: float

    @ShapeInterface.register
    class SquareConfig(BaseModel):
        shape_type: Literal["square"]
        side: float

Creating and Using Configurations
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

> A usage example creating and using plugin configurations is shown below.

Reviewer: The example below only shows how to use/instantiate the plugin. Also i think it needs a bit of elaboration what the "plugin" in this example is. I.e. how to instantiate plugins of the Shape base class?

.. code-block:: python

    # Direct instantiation using the plugin model
    Shape = ShapeInterface.plugin_model
    circle = Shape(
        shape={"shape_type": "circle", "radius": 5.0},

>         color="red"

Reviewer: This seems inconsistent to me. Why not 

```
circle = Shape(
        shape={"shape_type": "circle", "radius": 5.0,  "color": "red"}

)
```

or

```
circle = Shape(
        type= "circle",
        radius = 5.0,
        color="red"
)
```

otherwise it is a weird mix between kwargs and dictionary.

    )

    # Using the create class method
    square = ShapeInterface.create(
        shape={"shape_type": "square", "side": 3.0},
        color="blue"
    )

    # Accessing plugin-specific fields
    assert circle.shape.radius == 5.0
    assert square.shape.side == 3.0

Runtime Extension
~~~~~~~~~~~~~~~~~

The plugin system supports runtime registration of new plugin configurations.

> However, this requires rebuilding the Type union which is done by the create method.

Reviewer: > rebuilding the type union 

Ok so my understanding from this that the type union is basically the entirety of the plugin types of a specific class, e.g. for turbulence it would hould k-epsilon, k-omega as a type union. I think this needs to spelled out explicitly somewhere.

Author: done in the overview

.. code-block:: python

    # Define new plugin configurations
    class TriangleConfig(BaseModel):
        shape_type: Literal["triangle"]
        base: float
        height: float

    class EllipseConfig(BaseModel):
        shape_type: Literal["ellipse"]
        major_axis: float
        minor_axis: float

    # Register at runtime
    ShapeInterface.register(TriangleConfig)
    ShapeInterface.register(EllipseConfig)

    # Use new configurations immediately
    triangle = ShapeInterface.create(
        shape={"shape_type": "triangle", "base": 4.0, "height": 3.0},
        color="green"

>     )

Reviewer: How is this example different then the previous one? It basically shows how to do it without the  @ShapeInterface.register decorator? This seems out of place here

Register multiple classes
~~~~~~~~~~~~~~~~~~~~~~~~~

> `AnimalInterface` will be stored in a central registry to show the availability of all registered plugins.

Reviewer: Why is it animals now instead of shapes?

.. code-block:: python

>     @PluginSystem.register(discriminator_variable="animal", discriminator="animal_type")

Reviewer: >  discriminator_variable="animal", discriminator="animal_type"

i still fail to understand what these are for?

        animal={"animal_type": "dog", "breed": "Golden Retriever", "is_trained": True},
        name="Buddy",
        age=3
    )

JSON Schema Generation
~~~~~~~~~~~~~~~~~~~~~~

The dynamically generated plugin models support JSON schema generation for validation and documentation purposes.
Json schemas show all possible configuration of all the available plugins of each base class and therefore enable the creation of user interfaces or config file validators.

.. code-block:: python

    # Get JSON schema for validation and documentation
    Shape = ShapeInterface.plugin_model
    schema = Shape.model_json_schema()

    # Schema includes discriminator mapping
    discriminator = schema["properties"]["shape"]["discriminator"]
    print(discriminator["propertyName"])  # "shape_type"
    print(discriminator["mapping"])       # {"circle": "...", "square": "..."}

Registry Management
~~~~~~~~~~~~~~~~~~~

The PluginSystem class provides utility methods to manage and inspect the plugin registries.
The following example demonstrates how to list registered plugins, retrieve specific registries, and remove plugins.

.. code-block:: python

    # List all plugin families

>     all_plugins = PluginSystem.list_plugins()

Reviewer: so the `families` could basically be turbulenceModels, viscosityModel, chemistryModel ...

Author: yes

Overview
--------

The plugin system allows the configuration of the used models via configuration files and is a runtime-extensible configuration system built on Pydantic discriminated unions and a registry pattern.

.. code-block:: cpp

    // RASModel would be discriminator
    // a tags / type field in the config file
    // that selects the right turbulence model
    RASModel        kOmegaSST;

    turbulence      on;

    printCoeffs     on;

A discriminated union selects the right config class based on a type/tag field (e.g., "RASModel": "kOmegaSST" vs "RASModel": "kEpsilon"), and each base class has its own registry of registered child implementations.
All plugins and all models are registered in a central registry to enable easy access and management for UI, validation purposes or generative AI.
So, the user can retrieve all available plugins, like turbulence models, boundary conditions, etc.  their configuration options and is able to validate them.
Additionally, the plugin system supports the generation of JSON schemas for documentation and validation purposes.

Discrimated Unions in Pydantic
------------------------------

> The discriminated union work in pydantic similar are given in the following simple example.

Reviewer: This not a sentence.
